### ISSUE TITLE
Alive documents increase unlimited?

### ISSUE THREAD

**Issue by razvanphp** [arangodb/arangodb#1596]
I think this is a bug, but also maybe I miss something in how the rev works, we have an edge collection with around 300k entries now, that is updated continuously from an arango job.

The edge gets huge, and this is how the figures look like:
<img width="551" alt="screen shot 2015-12-07 at 15 31 58" src="https://cloud.githubusercontent.com/assets/4599319/11629302/b47d32fc-9cf7-11e5-93ad-c1417a3e577a.png">

Why does it keep so many alive versions? Can I prevent it from doing that? The data is not that big, for sure. How can I force some "forget all revisions"?

This edge has only 1 index bucket, but that should have nothing to do with it.

Thank you!

**Comment by jsteemann** [comment]
How many documents/edges are in that collection at the moment? A `collection.count()` should show this. Is that number much deviating from the number shown for `Alive`?

**Comment by razvanphp** [comment]
Yes, I already said, 330.000 items vs 115.000.000 alive.

```
arangosh [_system]> db.sorter_api_at_product_category_sort.count();
355861

arangosh [_system]> db.sorter_api_at_product_category_sort.figures();
{
  "alive" : {
    "count" : 115439889,
    "size" : 23850362984
  },
  "dead" : {
    "count" : 608234,
    "size" : 125399232,
    "deletion" : 425001
  },
  "datafiles" : {
    "count" : 531,
    "fileSize" : 23964793264
  },
  "journals" : {
    "count" : 1,
    "fileSize" : 33554432
  },
  "compactors" : {
    "count" : 0,
    "fileSize" : 0
  },
  "shapefiles" : {
    "count" : 0,
    "fileSize" : 0
  },
  "shapes" : {
    "count" : 321,
    "size" : 71256
  },
  "attributes" : {
    "count" : 12,
    "size" : 624
  },
  "indexes" : {
    "count" : 2,
    "size" : 65936560
  },
  "lastTick" : "2386922340267405",
  "uncollectedLogfileEntries" : 0
}

arangosh [_system]>
```

**Comment by jsteemann** [comment]
Sorry, I overlooked the `300k`. 
115.000.000 alive documents seems way too much. When you say the collection is updated continuously, does that mean you remove and insert documents or really update or replace them? Which method do you use for that (i.e. `update()`, `replace()`, AQL...?). Thanks.

**Comment by razvanphp** [comment]
Actually this might be the reason, we use the graph module, something like this:

```
var graphModule = require('org/arangodb/general-graph'),
    db = require('org/arangodb').db,
    categoryGraph = graphModule._graph(collections.GRAPH_PRODUCT_CATEGORY);

var edge = db._collection(collections.EDGE_PRODUCT_CATEGORY).firstExample({
    _from: collections.PRODUCTS + '/' + productId,
    _to: collections.CATEGORIES + '/' + cat
});

if (edge) {
    categoryGraph[collections.EDGE_PRODUCT_CATEGORY].update(edge, {
        updated_at: timeOfUpdate
    });
} else {
    categoryGraph[collections.EDGE_PRODUCT_CATEGORY].save(
            collections.PRODUCTS + '/' + productId,
            collections.CATEGORIES + '/' + cat,
            {
                updated_at: timeOfUpdate
            }
    );
}
```

I would bet most of the time it gets into the update case, since the edge exists already.

**Comment by jsteemann** [comment]
Ok thanks, will try to reproduce this.

**Comment by razvanphp** [comment]
Thank you, in the meanwhile is there anything I can do to make it drop those revisions? If I restart it, it can't load them anymore because there is not enough RAM.

@pluma as a suggestion, can you please change all values of `Size` from the figures tab shown in the screenshot above, so that the right unit is calculated for each row, z.B. MB, GB, TB, since our data is often above 1000 MB and it gets harder to spot anomalies like the one in this ticket.

**Comment by pluma** [comment]
@razvanphp make sense. I've created a separate ticket for that as it's unrelated: #1597

**Comment by jsteemann** [comment]
It seems that at least after a `truncate()` operation some of the alive documents remain in the datafiles.
After a restart, the collection gets cleaned up properly and the stats will get in order again, but that requires loading the entire collection of course.

Is the compaction enabled for the collection in question? `db.sorter_api_at_product_category_sort.properties();` should show a `doCompact` value of `true`.

**Comment by razvanphp** [comment]
Yes it is. I also don't think we do any truncate on it...

```
arangosh [_system]> db.sorter_api_at_product_category_sort.properties();
{
  "doCompact" : true,
  "journalSize" : 33554432,
  "isSystem" : false,
  "isVolatile" : false,
  "waitForSync" : false,
  "keyOptions" : {
    "type" : "traditional",
    "allowUserKeys" : true
  },
  "indexBuckets" : 1
}

arangosh [_system]>
```

**Comment by jsteemann** [comment]
Not sure if this is related: 
I found that compaction can be blocked by JavaScript objects still pointing to documents in a collection.
For example, if you use a server-side job or Foxx action that fetches a document from a collection and stores it inside a JavaScript variable, this will prevent the datafile the document is contained in from being deleted. This is not a problem if documents are only stored in temporary JavaScript variables or local variables in functions that are left again. It can be a problem if documents are assigned to global JavaScript variables or instance variables if the objects are long-lived. In this case, the reference pointing from the JavaScript variable to the document will prevent the compaction from making progress. Compaction will still run, but it cannot remove old files because JavaScript variables are still pointing into them.
This can also happen when only local JavaScript variables are temporaries point to the documents but garbage collection hasn't run. Once garbage collection has run and no variables point to the documents, everything proceeds and compaction will start removing obsolete files.
Whenever I tried it, I saw the `compactor` attribute of the `figures()` command show a value of `1` in its `count` subattribute. That's the indication for the compaction not making progress. Looking at the `figures()` output you posted, it seems like that's not what's happening in your case.

**Comment by razvanphp** [comment]
Yesterday I used the huge arangodump to restore the data on a new machine. It took the alive documents with it, but after unloading and loading again the collection, all excess edges got into Dead, and they were cleaned after a few minutes as expected.

BTW: maybe I trigger another unexpected thing, I clicked unload, and then load again maybe too fast, and now it does not want to unload anymore. Every time I try it remains in "in the process of being unloaded". If I click on it, I can see the data instantly, so I don't think he tries to unload it anymore. Also the alive documents stay. The edge data files are 1.7 GB.

**Comment by jsteemann** [comment]
There are a few subtleties with unloading: if some script or code still has a reference to a document in the collection that is going to be unloaded, then the unload will not happen until the reference to the document goes out of scope or is garbage collected. The collection will only change its status from `loaded` to `in the process of being unloaded` and wait there until no references point to it. When the last reference is gone, the actual unload will happen. Another reason for a collection not physically unloading is that data for the collection is still present only in the write-ahead-log, but not in the collection's datafiles. If that happens, the collection will not physically unload until the collection data is transferred from the WAL into the collection's datafiles. The data will be transferred only for full WAL files, and as each is 32 MB by default, it may take a while until a WAL file gets full and the collection can proceed with its unloading.
The currently open WAL file is intentionally not flushed every time a collection is unloaded because that may be expensive.

I understand this is unintuitive because the message `in the process of being unloaded` suggests there will soon be progress, while the actual status message should rather be `waiting for references pointing to the collection to go out of scope and maybe also a WAL flush`. I hope this can be improved somehow, and even if only by making the UI or status messages more understandable. But ideally there'd be a way to force the unload somehow. I need to discuss this internally.

**Comment by jsteemann** [comment]
Regarding compaction: I have some work in progress regarding compaction. This should address some of its issues and provide better debuggability for cases when its unclear what the compaction status is.
I am currently working in some side branch, but I hope to get the branch ready until the end of the week.

**Comment by razvanphp** [comment]
Any updates regarding this one? Can you reproduce this or we need the new version with improved logging to find out? Maybe I can provide the Foxx app that does this.

It gets a little bit tricky, since this eats up all RAM of the server, so I guess it starts to "pseudo-swap", unloading the datafiles from memory.

Luckily it seems that unloading and loading the collection sometimes moves all documents into `Dead` and they get garbage collected.

**Comment by jsteemann** [comment]
Yes, unloading and re-loading the collection will re-assess the number of alive/dead documents. This should enable compaction to remove the dead documents.

The status of this issue is:

* there's no additional logging as we currently don't have a good mechanism to turn on/off logging selectively just for some functions or resources such as collections. However, compaction progress and non-progress are now reported by the `figures()` method for a collection and the web interface. This can be used for debugging compaction problems. This was a substantial change including some API changes, so it is present in the 2.8 and devel branches only.
* I can confirm the compaction can stall when an unload request for the collection is made but some later activity happens that requires the collection and implicitly reloads it before the unload request is actually carried out. The fix is included in the 2.8 and devel branches at the moment.

It's unclear to me whether the latter is the root cause for the compaction stall that seems to be happening on your server, but I doubt it. Even if compaction does not make any progress, `alive` size should not get much higher if documents get only updated. 
So this should be further investigated. If you could provide the Foxx app (if it's doing more with the collection than shown in the above code snippet) and the collection definitions (indexes etc.) that would be great, because it may help us finding the root cause.

**Comment by jsteemann** [comment]
Potential improvements for this are included in 2.7.4. That version has a garbage collection modification that will collect document references held by V8 in all V8 contexts eventually, which may have been a source for the compaction being stalled. It also contains a fix for the compaction not making any progress when an unload of a collection is attempted and the collection is used again before the unloading starts. That version also allows querying the compaction status via a collection's `figures()` command, and its output is also shown in the web interface (by the *i* icon for a collection).

**Comment by jsteemann** [comment]
Datafile statistics have been re-implemented in 2.7.5 and 2.8.0-beta8. That means the number of alive (and dead) documents should be much more accurate in the versions, and the number of datafiles should not grow substantially in the long run, provided the number of live documents stays about constant.

**Comment by razvanphp** [comment]
I can confirm, the compaction seems to work now, both in 2.8.0-beta8 and 2.7.5, we have them both installed on 2 separate machines.

Thank you!
